File: DisplayServer/framebuffer.py
from font import Font
from bitmap import Bitmap

DISPLAY_WIDTH = 160
DISPLAY_HEIGHT = 24

DATA_SIZE = DISPLAY_WIDTH*DISPLAY_HEIGHT/8

OVERSCROLL = 10

class TextArea:
	x = 0
	y = 0
	w = 0
	h = 0

	def __init__(self):
		self.clear()

	def set(self, x, y, w, h):
		self.x, self.y, self.w, self.h = x, y, w, h

	def clear(self):
		self.w = 0
		self.h = 0

	def isClear(self):
		return (self.w == 0 and self.h == 0)

	def isInside(self, x, y):
		return (x >= self.x and y >= self.y and x < (self.x + self.w) and y < (self.y + self.h))

class FrameBuffer:
	_textarea = TextArea()

	def __init__(self):
		self._data = [0 for i in range(DATA_SIZE)]
		self.clearTextArea()
		self._textBmps = [None, None]
		self._texts = [u"", u""]
		self._textOffsets = [0, 0]
		self._font = Font(9)

	def setPixel(self, x, y, on):
		# range check
		if x < 0 or y < 0 or x >= DISPLAY_WIDTH or y >= DISPLAY_HEIGHT:
			return False

		bit  = (y * DISPLAY_WIDTH + x)
		byte = bit / 8
		bit %= 8

		if on:
			self._data[byte] |= (0x80 >> bit);
		else:
			self._data[byte] &= ~(0x80 >> bit);

		return True

	def setText(self, line, s):
		if line > 1:
			raise IndexError()

		if self._texts[line] != s:
			self._textBmps[line] = self._font.getBitmap(s)
			self._texts[line] = s

	def redrawText(self):
		# do not render anything if textarea is clear
		if self._textarea.isClear():
			return

		for line in range(2):
			bmp = self._textBmps[line]

			if not bmp:
				continue

			if self._textOffsets[line] >= 0:
				sx = self._textarea.x
			elif self._textOffsets[line] < (self._textarea.w - self._textBmps[line].width):
				sx = self._textarea.x + self._textarea.w - self._textBmps[line].width
			else:
				sx = self._textarea.x + self._textOffsets[line]

			sy = self._textarea.y + bmp.height * line

			for dx in range(bmp.width):
				for dy in range(bmp.height):
					x = sx+dx
					y = sy+dy

					if self._textarea.isInside(x, y):
						self.setPixel(x, y, bmp.isOn(dx, dy))

	def shiftText(self):
		for line in range(2):
			if not self._textBmps[line]:
				continue

			# check if scrolling is necessary
			if self._textBmps[line].width > self._textarea.w:
				self._textOffsets[line] -= 1

				# if text would vanish on the left, reset it
				if self._textOffsets[line] < (self._textarea.w - self._textBmps[line].width - OVERSCROLL):
					self._textOffsets[line] = OVERSCROLL

	def clear(self, on = False):
		for i in range(DATA_SIZE):
			if on:
				self._data[i] = 0xFF
			else:
				self._data[i] = 0x00

	def updateFromString(self, s):
		if len(s) >= DATA_SIZE:
			self._data = [ord(s[i]) for i in range(DATA_SIZE)]
		else:
			raise IndexError()

	def drawBitmap(self, x, y, bitmap):
		for dx in range(bitmap.width):
			for dy in range(bitmap.height):
				self.setPixel(x+dx, y+dy, bitmap.isOn(dx, dy))

	def setTextArea(self, x, y, w, h):
		self._textarea.set(x, y, w, h)

	def clearTextArea(self):
		self._textarea.clear()

	def serialize(self):
		s = "\x1b\x02"
		for byte in self._data:
			if byte == 0x1B:
				# this is the escape byte, which must be repeated in data
				s += chr(byte)
			s += chr(byte)

		s += "\x1b\x03"
		return s

File: DisplayServer/src/Framebuffer.cpp
#include <sstream>

#include "Exception.h"
#include "Logger.h"

#include "Framebuffer.h"

Framebuffer::Framebuffer(FT_Library *ftlib)
	: Bitmap(Framebuffer::WIDTH, Framebuffer::HEIGHT)
{
	this->clear(false);

	m_textBitmaps[0] = new Bitmap(0, 0);
	m_textBitmaps[1] = new Bitmap(0, 0);
	m_font = new Font(ftlib, "/usr/share/fonts/dejavu/DejaVuSans.ttf", 10);
}

Framebuffer::~Framebuffer()
{
	delete m_textBitmaps[0];
	delete m_textBitmaps[1];
	delete m_font;
}

void Framebuffer::serialize(std::string *buffer) const
{
	std::ostringstream oss;

	oss.write("\x1B\x02", 2); // start sequence

	for(Bitmap::BitmapData::size_type i = 0; i < (WIDTH*HEIGHT/8); i++) {
		if(m_data[i] == 0x1B) {
			// duplicate escape character
			oss.write("\x1B", 1);
		}

		oss.write(reinterpret_cast<const char*>(&(m_data[i])), 1);
	}

	oss.write("\x1B\x03", 2); // end sequence

	*buffer = oss.str();
}

void Framebuffer::drawText(int x, int y, const std::wstring &text)
{
	Bitmap textBitmap(0, 0);
	textBitmap.clear(false);
	m_font->renderText(text, &textBitmap);

	this->blit(textBitmap, x, y);
}

void Framebuffer::setTextArea(int x, int y, unsigned w, unsigned h)
{
	m_textArea.x = x;
	m_textArea.y = y;
	m_textArea.w = w;
	m_textArea.h = h;
}

void Framebuffer::setText(unsigned line, const std::wstring &text)
{
	if(line > 1) {
		throw RangeException("Framebuffer", "Line is out of range", line, 1);
	}

	if(m_texts[line] != text) {
		m_font->renderText(text, m_textBitmaps[line]);
		m_texts[line] = text;
		m_textOffsets[line] = OVERSCROLL;
	}
}

void Framebuffer::redrawText(void)
{
	if(m_textArea.isClear()) {
		return;
	}

	// clear the text area
	for(int dx = 0; dx < m_textArea.w; dx++) {
		for(int dy = 0; dy < m_textArea.w; dy++) {
			this->setPixel(m_textArea.x + dx, m_textArea.y + dy, 0);
		}
	}

	// render the text
	for(int line = 0; line < 2; line++) {
		Bitmap *bmp = m_textBitmaps[line];

		int sx;
		if(m_textOffsets[line] >= 0) {
			sx = 0;
		} else if(m_textOffsets[line] < ((int)m_textArea.w - (int)bmp->getWidth())) {
			sx = (int)m_textArea.w - (int)bmp->getWidth();
		} else {
			sx = m_textOffsets[line];
		}

		int miny = LINEHEIGHT * line;
		int maxy = LINEHEIGHT * (line + 1);

		if(maxy > m_textArea.h) {
			maxy = m_textArea.h;
		}

		for(int dx = 0; dx < m_textArea.w; dx++) {
			for(int dy = miny; dy < maxy; dy++) {
				int x = m_textArea.x + dx;
				int y = m_textArea.y + dy;

				this->setPixel(x, y, bmp->getPixel(dx - sx, dy - miny));
			}
		}
	}
}

void Framebuffer::shiftText(void)
{
	for(int line = 0; line < 2; line++) {
		Bitmap *bmp = m_textBitmaps[line];

		if(bmp->getWidth() > m_textArea.w) {
			m_textOffsets[line]--;

			if(m_textOffsets[line] < ((int)m_textArea.w - (int)bmp->getWidth() - OVERSCROLL)) {
				m_textOffsets[line] = OVERSCROLL;
			}
		}
	}
}
